File: examples/scrape-github-trending/rest/nodejs/scrape-github-trending.mjs
// Scrapes GitHub trending repositories using BQL.
//
// Run: node scrape-github-trending.mjs

const query = `mutation ScrapeGitHubTrending {
  goto(url: "https://github.com/trending", waitUntil: networkIdle) {
    status
  }
  waitForSelector(selector: "article.Box-row", timeout: 15000) {
    time
  }
  repos: mapSelector(selector: "article.Box-row") {
    name: mapSelector(selector: "h2 a") { innerText }
    description: mapSelector(selector: "p") { innerText }
    language: mapSelector(selector: "[itemprop=programmingLanguage]") { innerText }
    stars: mapSelector(selector: "a[href*=stargazers]") { innerText }
    forks: mapSelector(selector: "a[href*=forks]") { innerText }
    todayStars: mapSelector(selector: "span.d-inline-block.float-sm-right") { innerText }
  }
}`;

const response = await fetch(
  'https://production-sfo.browserless.io/chromium/bql?token=YOUR_API_TOKEN_HERE',
  {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ query, variables: {}, operationName: 'ScrapeGitHubTrending' }),
  }
);

const { data } = await response.json();
const repos = data.repos.map((r) => ({
  name: r.name?.[0]?.innerText?.trim() ?? '',
  description: r.description?.[0]?.innerText ?? '',
  language: r.language?.[0]?.innerText ?? '',
  stars: r.stars?.[0]?.innerText?.trim() ?? '',
}));
console.log(JSON.stringify(repos, null, 2));

File: examples/scrape-github-trending/rest/java/ScrapeGitHubTrending.java
// Scrapes GitHub trending repositories using BQL.
//
// Run: javac ScrapeGitHubTrending.java && java ScrapeGitHubTrending

import java.net.URI;
import java.net.http.*;

public class ScrapeGitHubTrending {
    public static void main(String[] args) throws Exception {
        String token = "YOUR_API_TOKEN_HERE";
        String endpoint = "https://production-sfo.browserless.io/chromium/bql?token=" + token;

        String query = "mutation ScrapeGitHubTrending { goto(url: \\"https://github.com/trending\\", waitUntil: networkIdle) { status } waitForSelector(selector: \\"article.Box-row\\", timeout: 15000) { time } repos: mapSelector(selector: \\"article.Box-row\\") { name: mapSelector(selector: \\"h2 a\\") { innerText } description: mapSelector(selector: \\"p\\") { innerText } language: mapSelector(selector: \\"[itemprop=programmingLanguage]\\") { innerText } stars: mapSelector(selector: \\"a[href*=stargazers]\\") { innerText } forks: mapSelector(selector: \\"a[href*=forks]\\") { innerText } todayStars: mapSelector(selector: \\"span.d-inline-block.float-sm-right\\") { innerText } } }";

        String payload = "{\"query\": \"" + query + "\", \"variables\": {}, \"operationName\": \"ScrapeGitHubTrending\"}";

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(payload))
            .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        System.out.println(response.body());
    }
}
